appservice/multiplexer.py

```diff
@@ -153,104 +69,166 @@ def new_session_podman(self, sessionid):
+                    await upstream_resp.prepare(upstream_req)
+                    async for data, _ in downstream_response.content.iter_chunks():
+                        await upstream_resp.write(data)
+                    await upstream_resp.write_eof()
+                    return upstream_resp
+                else:
+                    upstream_resp = web.Response(
+                        status=downstream_response.status,
+                        reason=downstream_response.reason,
+                        headers=h,
+                        body=await downstream_response.content.read(),
+                    )
+
+                return upstream_resp
+
+
+async def init_sessions():
+    global SESSIONS
+    sessions = await REDIS.get('sessions')
+    if sessions is None:
+        SESSIONS = {}
+    else:
+        try:
+            SESSIONS = json.loads(sessions)
+        except json.decoder.JSONDecodeError:
+            SESSIONS = {}
 
+    logger.debug('initial sessions: %s', SESSIONS)
 
-def start_nginx():
-    proc = subprocess.Popen(['nginx'])
 
-    # wait for nginx to start up
-    connection = http.client.HTTPConnection('localhost:8080')
-    for _ in range(10):
+async def watch_redis(channel):
+    global SESSIONS
+    while True:
         try:
-            connection.connect()
-            break
-        except OSError:
-            time.sleep(0.2)
-    else:
-        raise TimeoutError('timed out waiting for nginx to start up')
-    return proc
+            async with async_timeout.timeout(1):
+                message = await channel.get_message(ignore_subscribe_messages=True)
+                if message is not None and message['channel'] == b'sessions':
+                    logger.debug("got redis sessions update: %s", message['data'])
+                    try:
+                        SESSIONS = json.loads(message['data'].decode())
+                    except json.decoder.JSONDecodeError as e:
+                        logger.warning("invalid JSON, starting without sessions: %s", e)
+                        SESSIONS = {}
+
+                await asyncio.sleep(0.01)
+        except asyncio.TimeoutError:
+            pass
+
+
+async def main():
+    async with Handlers() as handlers:
+        pubsub = REDIS.pubsub()
+        await pubsub.subscribe('sessions')
+        asyncio.create_task(watch_redis(pubsub))
+        await init_sessions()
+
+        app = web.Application()
+        app.router.add_route('GET', f'{config.ROUTE_API}/ping', handlers.handle_ping)
+        app.router.add_route('GET', f'{config.ROUTE_API}/sessions/new', handlers.handle_session_new)
+        app.router.add_route('*', f'{config.ROUTE_WSS}/sessions/{{sessionid}}/{{path:.*}}', handlers.handle_session_id)
+
+        runner = web.AppRunner(app, auto_decompress=False)
+        await runner.setup()
+        await web.TCPSite(runner, port=8080).start()
+        while True:
+            await asyncio.sleep(60)
```

Comment:
We can't away on the .. TCPSite? This feels a little weird.

Reply:
Indeed https://docs.aiohttp.org/en/stable/web_quickstart.html#run-a-simple-web-server shows how to run the app in a much more tasteful way. However, I tried to replace this while loop with `web.run_app(app)`, and it just barfs out with `RuntimeError: Cannot run the event loop while another loop is running`. Possibly because I already added the watch_redis task, and thus it already has a main loop? Do you have a better idea here?

Reply:
Hmm, it's actually what https://docs.aiohttp.org/en/stable/web_advanced.html#aiohttp-web-app-runners says to do..

Reply:
Actually, https://docs.aiohttp.org/en/stable/web_advanced.html#background-tasks describes exactly our use case! I tried the diff below, but alas, still `RuntimeError: Cannot run the event loop while another loop is running`. I've tried to wire this up differently in about 10 different iterations, I just can't make it work :cry: 

```diff
--- appservice/multiplexer.py
+++ appservice/multiplexer.py
@@ -227,10 +227,10 @@ async def main():
     async with Handlers() as handlers:
         pubsub = REDIS.pubsub()
         await pubsub.subscribe('sessions')
-        asyncio.create_task(watch_redis(pubsub))
         await init_sessions()
 
         app = web.Application()
+        app.on_startup.append(watch_redis(pubsub))
         app.router.add_route('GET', f'{config.ROUTE_API}/ping', handlers.handle_ping)
         app.router.add_route('GET', f'{config.ROUTE_API}/sessions/new', handlers.handle_session_new)
         app.router.add_route('GET', f'{config.ROUTE_API}/sessions/{{sessionid}}/status', handlers.handle_session_status)
@@ -239,8 +239,7 @@ async def main():
         runner = web.AppRunner(app, auto_decompress=False)
         await runner.setup()
         await web.TCPSite(runner, port=8080).start()
-        while True:
-            await asyncio.sleep(60)
+        web.run_app(app)
 
 
 if __name__ == '__main__':
```

Reply:
This looks much nicer with starlette, and I can't figure out something better with aiohttp. So closing this for now.